feat(core): introduce the reactive linkedSignal

pkozlowski-opensource opened the pull request:
This change introduces the new reactive primitive: linkedSignal.

A linkedSignal represents state (hence the signal in the name) that is reset based on the provided computation. Conceptually it is a state that is maintained / valid only in the context of another source signal (context is deteremined by a computation).

dostarora97 commented:
Sorry if I’m missing something obvious here.
But how is this different from [`computed()` signal](https://angular.dev/guide/signals#computed-signals) ?

Armenvardanyan95 commented:
> Sorry if I’m missing something obvious here. But how is this different from [`computed()` signal](https://angular.dev/guide/signals#computed-signals) ?

This will be writable. Computed signals are only readable, so they completely depend on their tracked signals. With this, you can set your own values, but it will be overwritten if the source signal changes its value

alxhub commented on goldens/public-api/core/primitives/signals/index.api.md:
````diff
@@ -4,6 +4,11 @@
 ```ts
 
+// @public
+export type ComputedGetter<T> = (() => T) & {
````
It's a draft PR, but note that primitives changes cannot be in the same PR as Angular changes.

pkozlowski-opensource replied:
Oh, right, thnx - will split it before marking anything for merge.

pkozlowski-opensource replied:
This is split in commits / PR now.
Primitives PR: https://github.com/angular/angular/pull/58228

alxhub commented on packages/core/src/render3/reactivity/linked_signal.ts:
```diff
@@ -0,0 +1,107 @@
+
+        const previous =
+          previousSource !== undefined && node !== undefined
+            ? {source: previousSource, value: node.previousValue!}
```
Is `node.value` not the previous value that we care about here? Do we need `.previousValue` at all?

pkozlowski-opensource replied:
`node.value` will be set to the `COMPUTING` symbol during the computation.

alxhub commented on packages/core/src/render3/reactivity/linked_signal.ts:
```diff
@@ -0,0 +1,107 @@
+  const useShorthand = typeof optionsOrComputation === 'function';
+
+  let previousSource: S | undefined;
+  let node: LinkedSignalNode<D> | undefined;
```
`node` will always be defined, and we can probably type it that way.

pkozlowski-opensource replied:
Done

alxhub commented on packages/core/primitives/signals/src/computed.ts:
```diff
@@ -50,11 +52,18 @@ export type ComputedGetter<T> = (() => T) & {
 
   const computed = () => {
+    if (capturePreviousValue) {
+      node.previousValue = node.value;
```
Capturing the previous value during _reads_ was a bit surprising to me. Digging into it a bit, I think there's actually a problem here: nothing guarantees the linkedSignal will be read, so the previous value may not update.

pkozlowski-opensource replied:
This should be fine I think as an explicitly set signal will have a value. And we are capturing the value _just_ before the recompute - where we actually need the previous value.

pkozlowski-opensource replied:
Changed the way previous values are handled, moving the capture of them to the assign phase.

alxhub commented on packages/core/src/render3/reactivity/linked_signal.ts:
```diff
@@ -0,0 +1,116 @@
+        const newChoice = untracked(() => options.computation(source, previous));
+        previousSource = source;
+
+        if (!node.equal(node.previousValue!, newChoice)) {
```
Ugh, this is so complex. I don't think we should be recomparing through equality :-/ that could be expensive (deepEquals comes to mind).

The more I look at this, the more I think a dedicated node type with custom logic would be far more straightforward.

alxhub replied:
Yeah, thinking about this more, I'm pretty convinced now. We would be able to store `source` and `computation` separately, and always call `computation` with the right arguments for previous value. Then the shorthand form becomes a node with the callback for `source` and the identity function for `computation`. I think it'd be more readable as well.

pkozlowski-opensource replied:
I took a stab at it in https://github.com/angular/angular/pull/58189/commits/df4f7e354767c714d4fa91b4881e924794eaa6a3 

it does make logic easier to follow at the expense of duplicating code from the `computed`. But I do tend to agree with you that it is worth it. PTAL.

pkozlowski-opensource replied:
Pair-reviewed offline, we are keeping the version with a dedicated reactive node (latest iteration).

alxhub commented on packages/core/test/signals/linked_signal_spec.ts:
```diff
@@ -0,0 +1,227 @@
+
+  it('should expose previous source in the computation function', () => {
+    const options = signal(['apple', 'banana', 'fig']);
+    // THINK: we need to specify the D generic here which is slightly inconvenient
```
I poked at this a bit today. Ultimately its due to the object form of `previous` - TypeScript is unable to infer `D` when it's used inside an object type which contributes to the function's return type. A form that does work is:
```ts
computation: (options: S, prevOptions?: S, prevValue?: D)
```
but of course, that has all the consequences of positional arguments. Given this, I think we can live with the generic being required for now.

pkozlowski-opensource replied:
Ack.

alxhub commented on packages/core/test/signals/linked_signal_spec.ts:
```diff
@@ -0,0 +1,227 @@
+    }
+  });
+
+  it('should capture linked signal dependencies even if the value is set before the initial read', () => {
```
Another useful test would be a similar one as this but for `update`, which verifies that the initial value is computed prior to running the updater function.

pkozlowski-opensource replied:
Added

pkozlowski-opensource committed:
feat(core): introduce the reactive linkedSignal

This change introduces the new reactive primitive: linkedSignal.

A linkedSignal represents state (hence the signal in the name)
that is reset based on the provided computation. Conceptually
it is a state that is maintained / valid only in the context of
another source signal (context is deteremined by a computation).

Closes #55673

chrisrocco reviewed (commented):
I believe it's possible to construct this API using existing primitives. Was this approach considered, and/or what are the benefits of creating a new subclass of ReactiveNode instead? Performance? DevTools?

Something like this (plus the other features) -

```ts
function linkedSignal(src: Signal) {
  const wrapped = () => signal(src());
  const getter = () => wrapped()[0]();
  const setter = (next) => wrapped()[1](next);
  getter.set = setter;
  return getter;
}
```

alxhub reviewed (approved):
Reviewed-for: public-api

pkozlowski-opensource commented:
@chrisrocco there are some complications with implementing this purely in the user-land: mostly around previous values and equality. Will follow up with you offline.

alxhub reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api, fw-core

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

devversion commented:
This PR was merged into the repository by commit 8311f00faaf282d1a5b1ddca29247a2fba94a692.

The changes were merged into the following branches: main

Ookamini95 commented:
Not the best programmer, but i belive "linked" would keep close to the style of "signal" and "computed". Nothing wrong with longer names, but if you really want to go for it why not a more intuitive name like "writableComputed"? What is the reasoning behind "linkedSignal" naming convention?
